### PR TITLE
Rewrite build system for new versions and Heroku stacks

### LINKED ISSUE #2: Upgrade to 4.0.4

### LINKED ISSUE #7: New Heroku-16 stack available
Not sure if we actually need to change anything for this or not

### LINKED ISSUE #13: No heroku-16 build
I have:

```
$  heroku buildpacks
1. https://github.com/heroku/heroku-buildpack-apt
2. https://github.com/mojodna/heroku-buildpack-jemalloc.git#v3.6.0
3. heroku/ruby
```

When I deploy, the build phase crashes:

```
[...]
remote: -----> jemalloc app detected
remote: -----> Vendoring binaries
remote:        Fetching https://s3.amazonaws.com/mojodna-heroku/heroku-16/jemalloc-3.6.0-1.tar.gz
remote: 
remote: gzip: stdin: not in gzip format
remote: tar: Child returned status 1
remote: tar: Error is not recoverable: exiting now
remote:  !     Push rejected, failed to compile jemalloc app.
remote: 
remote:  !     Push failed
[...]
```

This file doesn't exist: https://s3.amazonaws.com/mojodna-heroku/heroku-16/jemalloc-3.6.0-1.tar.gz

Everything is fine if I don't pin to `v3.6.0`.

Is there a reason 3.6.0 is still shown as a fallback option? Is it still supported?

### PR DESCRIPTION
**Pull request by gaffneyc**
This allows new versions to be built for uploading by running
`make VERSION=5.0.1` rather than updating the hard coded version in
multiple places. I've also added support for building against either the
legacy cedar-14 architecture or heroku-16. While versions built on
cedar-14 appear to run on heroku-16, the newer stack should also mean a
newer and better tool chain.

I've removed the Dockerfile and switched to using Heroku's docker images
directly as this removes the need for cleaning them up afterwards. It
possibly makes debugging harder but having build.sh available should
avoid some of that.

Fixes #2
Fixes #7
Fixes #13

### CODE REVIEW

**Comment by gaffneyc** [comment]
This is the first part of some work I've been doing today. Started as a review of what was here and... kind of spiraled. I'm also looking at making changes to `bin/compile` but wanted to discuss them before getting too deep into it.

The goal here was to make it easy to create builds for new versions of jemalloc. The next step that I'm working on is being able to use those builds and easily switch between them. From a high level this codebase is two things: heroku buildpack for downloading compiled binaries and tooling to build those binaries.

It should be possible to quickly and easily make binaries available based on version without needing to update the buildpack. I'm thinking we can use a set of releases (one per stack) that hosts the binaries for each version of jemalloc ([I have a proof of concept on my fork](github.com/gaffneyc/heroku-buildpack-jemalloc/releases)).

For switching between versions, I've been able to have a `JEMALLOC_VERSION` env variable that will compile different versions when a new slug is compiled. I am also considering a `JEMALLOC_ENABLED` flag that will automate the process of setting `LD_PRELOAD`.

**Comment by gaffneyc** [comment]
In `bin/compile` is it necessary to set `LD_LIBRARY_PATH`, `LIBRARY_PATH`, `PKG_CONFIG`, `CPPPATH`, and `CPATH`? I believe those are only necessary if someone plans to link against jemalloc but only `LD_PRELOAD` is required for using the library. Removing them from `bin/compile` would reduce the file quite a bit.

**Comment by csuhta** [comment]
:+1: on the environment variables to control the build and if jemalloc should be used. That makes testing very easy too. 

Using `jemalloc.sh` is fairly awkward right now.

**Comment by gaffneyc** [comment]
I've fleshed out what I was thinking and it's sitting on [master of my fork](https://github.com/gaffneyc/heroku-buildpack-jemalloc). We're slowly testing it on a couple apps and already have it running on [Dead Man's Snitch](https://deadmanssnitch.com).

Going to clean up some of the fork specific changes and get another PR up for consideration.

**Comment by csuhta** [comment]
@gaffneyc I was trying out your folk, and I noticed the builds for jemalloc 3.6.0 do not include a `jemalloc-config` script in the archive

**Comment by gaffneyc** [comment]
@csuhta Interesting, I will take a look. I'm not modifying the sources at all so I'm guessing it's not part of the 3.x line. I'll see about getting a work around in place.

Going to take some time today to upstream my changes as well.

**Comment by csuhta** [comment]
@gaffneyc I created two bundles with working config scripts for 3.6.0:

https://img.scryfall.com/build/jemalloc/cedar-14/jemalloc-3.6.0.tar.bz2
https://img.scryfall.com/build/jemalloc/heroku-16/jemalloc-3.6.0.tar.bz2

**Comment by gaffneyc** [comment]
Was there a config change you needed to make for it to be generated or did you add the file directly?

**Comment by csuhta** [comment]
I manually created them

**Comment by gaffneyc** [comment]
Gotcha. It looks like jemalloc-config was added in 4.0.0. Is it worth support any versions older than 4.0.0 other than 3.6.0?

**Comment by gaffneyc** [comment]
It may be possible to hardcode it as /app/vendor/jemalloc/lib/libjemalloc.so as it's a symlink to the specific revision. Since slugs are compiled independently on each code change, there should only ever be one version installed in /app/vendor/jemalloc

**Comment by gaffneyc** [comment]
@csuhta I've pushed https://github.com/gaffneyc/heroku-buildpack-jemalloc/commit/6564f30baf04e57d3809dfd51a63fb006cd6b3d4 which should fix the issue with 3.6.0. I was able to remove the need for jemalloc-config entirely.

File: Makefile [hunk 1]
```diff
@@ -1,21 +1,33 @@
...
-src/jemalloc.tar.bz2:
+# Download missing source archives to ./src/
+src/jemalloc-%.tar.bz2:
 	mkdir -p $$(dirname $@)
-	curl -sL https://github.com/jemalloc/jemalloc/releases/download/5.0.0/jemalloc-5.0.0.tar.bz2 -o $@
+	curl -fsL https://github.com/jemalloc/jemalloc/releases/download/$*/jemalloc-$*.tar.bz2 -o $@
+
+.PHONY: cedar-14 heroku-16
+
+# Build for cedar stack (deprecated)
```
Comment (nateberkopec) [inline Makefile:16]: Cedar-14 isn't deprecated, can just remove this comment

**Comment by nateberkopec** [comment]
Thank you so much for this.

>  I am also considering a JEMALLOC_ENABLED flag that will automate the process of setting LD_PRELOAD.

I like this, but can leave for another day/PR.

**Comment by nateberkopec** [comment]
Maybe also for a different PR, but to use any of this properly, we'll have to change the way this line works in [bin/compile](https://github.com/mojodna/heroku-buildpack-jemalloc/blob/3f76484bc735d47b83388260102ef5425aaf64d3/bin/compile#L38), correct?

**Comment by gaffneyc** [comment]
I've been planning to create a PR for what's on the master branch of my fork but have been pressed for time this last week. Thinking that PR would supersede this one and probably #11.

**Comment by gaffneyc** [comment]
Yes, it will end up looking something like this: https://github.com/gaffneyc/heroku-buildpack-jemalloc/blob/6306663a60b49aadc2f2907141028f07029d59e7/bin/compile#L18 with releases configured like this: https://github.com/gaffneyc/heroku-buildpack-jemalloc/releases

**Comment by gaffneyc** [comment]
Took a couple minutes and submitted #18 which supersedes this one.

**Pull request closed without merging**